File: skyspark-scala/src/main/scala/com/github/dkanellis/skyspark/scala/api/algorithms/bnl/LocalSkylineCalculator.scala
package com.github.dkanellis.skyspark.scala.api.algorithms.bnl

import com.github.dkanellis.skyspark.scala.api.algorithms.Point
import org.apache.spark.rdd.RDD

/**
  * This is responsible for the Reduce task of the Division part of the MR-BNL algorithm.
  *
  * @param bnlAlgorithm The core BNL algorithm
  */
private[bnl] class LocalSkylineCalculator(private val bnlAlgorithm: BnlAlgorithm) extends Serializable {

  private[bnl] def computeLocalSkylines(flagPoints: RDD[(Flag, Point)]): RDD[(Flag, Point)] = {
    flagPoints
      .groupByKey
      .flatMapValues(bnlAlgorithm.computeSkylinesWithoutPreComparison)
  }
}

File: skyspark-java/src/main/java/com/akikanellis/skyspark/api/algorithms/bnl/LocalSkylineCalculator.java
package com.akikanellis.skyspark.api.algorithms.bnl;

import com.akikanellis.skyspark.api.algorithms.Point;
import org.apache.spark.api.java.JavaPairRDD;

import java.io.Serializable;

class LocalSkylineCalculator implements Serializable {
    private final BnlAlgorithm bnlAlgorithm;

    LocalSkylineCalculator(BnlAlgorithm bnlAlgorithm) { this.bnlAlgorithm = bnlAlgorithm; }

    JavaPairRDD<Flag, Point> computeLocalSkylines(JavaPairRDD<Flag, Point> flagPoints) {
        return flagPoints
                .groupByKey()
                .flatMapValues(bnlAlgorithm::computeSkylinesWithoutPreComparison);
    }
}
